### PR TITLE
Improve PHPStorm and PHPStan intergrations

### LINKED ISSUE #7073: Implement PHPStan lookup via injector / service container
## Feature Description

Breakout from https://github.com/google/web-stories-wp/pull/6906. Original comment https://github.com/google/web-stories-wp/pull/6906#discussion_r602178214. The amp team has done some great work on making the service container / injector model play nicely with PHPStan / static analytics. See can copy https://github.com/ampproject/amp-wp/tree/develop/tests/php/src/PhpStan and 
https://github.com/ampproject/amp-wp/commit/90a102be25e3403cc48dbe5e9f425c5fa4026bfd

This should make class lookups work as they did before working we used service containers. It means we can remove lines like this. 

https://github.com/google/web-stories-wp/blob/796e96c8cf40e295a6933be6f7bf4be0c14dd965/includes/namespace.php#L41-L45

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### LINKED ISSUE #7074: Implement PhpStorm advanced metadata file
## Feature Description

PHPStorm is a popular tool for developing PHP code. To implement development experience, adding a metadata file makes autocomplete work better. See original docs. 

From https://www.jetbrains.com/help/phpstorm/ide-advanced-metadata.html

> Besides built-in “code awareness” capabilities, PhpStorm relies on external knowledge of code, which comes in the form of PHP stubs and the special advanced metadata files.

> While stubs cover the Standard PHP Library components and common extensions, metadata files can be used for extending the PhpStorm functionality based on your own needs or project requirements. The basic metadata file, .phpstorm.meta.php, is bundled inside the PHP stubs package and located inside the meta folder.

Originally suggested in https://github.com/google/web-stories-wp/pull/6906#discussion_r602178214. We can copy AMP's example from https://github.com/ampproject/amp-wp/commit/4a337903c6e7e2b3f912451f9ab5b6de9d8ed565#diff-38ba3fecb1750ffc282c31a2b3271a035de8e76e2de71f63853ce248f7a1df83 and [.phpstorm.meta.php](https://github.com/ampproject/amp-wp/blob/develop/.phpstorm.meta.php)

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary
As #7073 and #7074 are heavily linked, I did them together. 

These make phpstan and phpstorm, be able to look up class via the inject. This makes auto complete and static analytics work again. 

See https://github.com/ampproject/amp-wp/pull/5320 https://github.com/ampproject/amp-wp/pull/5289 https://github.com/ampproject/amp-wp/pull/5410

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7073
Fixes #7074

### CODE REVIEW

File: includes/PhpStan/ServicesDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
+<?php
```
Comment (swissspidy) [inline includes/PhpStan/ServicesDynamicReturnTypeExtension.php]: PHPStan files should live in `tests/phpstan`, never in `includes/`

File: phpstan.neon.dist [hunk 2]
```diff
@@ -21,6 +29,7 @@ parameters:
     - includes/
   excludes_analyse:
     - includes/vendor/*
+    - includes/PhpStan/*
```
Comment (swissspidy) [inline phpstan.neon.dist]: PHPStan files should live in `tests/phpstan`, never in `includes/`

File: includes/PhpStan/ServicesDynamicReturnTypeExtension.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/PhpStan/ServicesDynamicReturnTypeExtension.php]: Fixed

File: phpstan.neon.dist (hunk 2 above)
Comment (spacedmonkey) [reply phpstan.neon.dist]: Fixed

**Review by swissspidy — CHANGES_REQUESTED** [review]
Some failing tests

**Comment by swissspidy** [comment]
Why was bce1352c78cc86f9a994f4b39e8c9fa0ffed4f7d needed here exactly? 👀

**Comment by spacedmonkey** [comment]
> Why was bce1352 needed here exactly? 👀

Fixes tests.

**Comment by swissspidy** [comment]
> > Why was [bce1352](https://github.com/google/web-stories-wp/commit/bce1352c78cc86f9a994f4b39e8c9fa0ffed4f7d) needed here exactly? 👀
> 
> Fixes tests.

I saw that the tests passed afterwards, but why? Why were only multisite tests broken? Why does `Services::get()` work everywhere else except there? If it's really not working, shouldn't there be a comment saying why this can't be changed?

Also, bce1352c78cc86f9a994f4b39e8c9fa0ffed4f7d changed quite a few lines when the test failures only complained about `database_upgrader`. So I would have only expected the `$database_upgrader = Services::get( 'database_upgrader' );` line to be changed, not others.

**Comment by spacedmonkey** [comment]
I was thinking of removing all the `Services::get` changes in this PR. They are not needed, but do test this change works. 

As for the changes in namespace, database_upgrader and story_post_type both are going to get refactored soon, so decided not to debug these issues.

**Comment by swissspidy** [comment]
> > > Why was [bce1352](https://github.com/google/web-stories-wp/commit/bce1352c78cc86f9a994f4b39e8c9fa0ffed4f7d) needed here exactly? 👀
> > 
> > 
> > Fixes tests.
> 
> I saw that the tests passed afterwards, but why? Why were only multisite tests broken? Why does `Services::get()` work everywhere else except there? If it's really not working, shouldn't there be a comment saying why this can't be changed?
> 
> Also, [bce1352](https://github.com/google/web-stories-wp/commit/bce1352c78cc86f9a994f4b39e8c9fa0ffed4f7d) changed quite a few lines when the test failures only complained about `database_upgrader`. So I would have only expected the `$database_upgrader = Services::get( 'database_upgrader' );` line to be changed, not others.

I debugged this myself now and left an inline comment on why it didn't work.

**Review by swissspidy — APPROVED** [review]

**Comment by schlessera** [comment]
@spacedmonkey, @swissspidy In the hopes this helps better understand where the above issues might be coming from, and why the tests could be flaky:

1. There is a fundamental difference between `Services::get( 'database_upgrader' )` and `Services::get( 'injector' )->make( Database_Upgrader::class )`. The first one retrieves the database upgrade from the service container, where it was put when the service was first instantiated. You'll always get that same one instance that was added to the service container. For the second one, you're asking the injector to instantiate the class for you. If the class was marked as being shared, this will result in the same thing than retrieving it from the service container, with the only difference that the service container is faster as it only needs a hashed lookup. However, if the class was not marked as being shared, the injector will instantiate a new instance every single time you `make()` it. Keep that in mind when dealing with the service container and injector, as it most of all has an impact on the tests and how state might or might not be persisted from one test to the other.

2. When you mark a service as being `Delayed` and let it only be instantiated on a certain action, it will not exist until that action is triggered. This seems to be why the database upgrade did break on multisite, because the `admin_init` action was not triggered. In that case, I'd consider this a bug in how the service is set up, as it's "timing" is configured in such a way that it can directly fail. I'd suggest moving the instantiation to an action that is always being triggered without fail, and then either add a `Conditional` that defines the exact requirements, or check them directly in the `register()` method.

**Comment by schlessera** [comment]
Ah, one other thing I failed to mention:
Consider the service container to be a compromise, a hack that let's you tap into the injection hierarchy from legacy code. The ideal case is a single `$injector->make()` call that instantiate your entire object hierarchy, with multiple `$injector->make()` calls (via an injected injector) being the next-best thing, and the use of the service container being an admission that the code is not yet fully ready for 100% injection. The service container, in the end, is nothing more than a glorified global access. So always think first if you can avoid it.

**Pull request merged**